apps/req.c

```diff
@@ -177,38 +176,36 @@ static void exts_cleanup(OPENSSL_STRING *x)
 static int duplicated(LHASH_OF(OPENSSL_STRING) *addexts, char *kv)
 {
     char *p;
+    size_t off;
 
     /* Check syntax. */
-    if (strchr(kv, '=') == NULL)
-        return 1;
-
     /* Skip leading whitespace, make a copy. */
     while (*kv && isspace(*kv))
         if (*++kv == '\0')
             return 1;
+    if ((p = strchr(kv, '=')) == NULL)
+        return 1;
+    off = p - kv;
     if ((kv = OPENSSL_strdup(kv)) == NULL)
         return -1;
 
     /* Skip trailing space before the equal sign. */
-    for (p = strchr(kv, '='); p > kv; --p)
-        if (p[-1] != ' ' && p[-1] != '\t')
+    for (p = kv + off; p > kv; --p)
+        if (!isspace(p[-1]))
             break;
     if (p == kv) {
         OPENSSL_free(kv);
         return 1;
     }
     *p = '\0';
 
-    /* Finally have a clean "key"; see if it's there. */
-    if (lh_OPENSSL_STRING_retrieve(addexts, (OPENSSL_STRING*)kv) != NULL) {
-        BIO_printf(bio_err, "Extension \"%s\" repeated\n", kv);
-        OPENSSL_free(kv);
-        return 1;
+    /* Finally have a clean "key"; see if it's there [by attempt to add it]. */
```

Comment:
This seems cleaner to me:
```
   /* "NULL is return on both error and success."  Aargh. */
    (void)lh_OPENSSL_STRING_insert(addexts, (OPENSSL_STRING*)kv);
    if (lh_OPENSSL_STRING_retrieve(addexts, (OPENSSL_STRING*)kv) == NULL) {
        OPENSSL_free(kv);
        return -1;
    }   
    return 0;
```
Either way, once you merge this I will add a positive test.

Reply:
That wouldn't work. Indeed, under normal circumstances insert *always* works, so that retrieve will always return non-NULL. And so 0 will always be returned and you won't detect duplicates. Thing to keep in mind is that insert actually replaces original pointer...

Reply:
Oh! Maybe you mean that removed retrieve would be put back? I mean suggestion was to remove retrieve followed by insert and replace it with insert *alone*. And you might mean to keep retrieve, so that it's still retrieve followed by insert. Is this it?

Reply:
No you understood me and I was wrong. :)

Reply:
All right, merged than, closing.